### PR TITLE
asserts,confdb: define confdb-control assertion

### PR DESCRIPTION
**Pull request by st3v3nmw**
Define the `confdb-control` assertion and add a feature flag. The bits on the API (delegating, revoking, etc) will be added later when the design discussions are finalized.

### CODE REVIEW

**Comment by st3v3nmw** [comment]
@MiguelPires

File: confdb/confdb_control.go
Comment (MiguelPires) [inline confdb/confdb_control.go:1]: I'd say lets put this in `registry/` for now so it's grouped up with the other code. Once I open rename, it will all get moved together

File: asserts/confdb_asserts.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		if !ok {
+			operator = &confdb.Operator{ID: operatorID}
+			cc.operators[operatorID] = operator
+		}
+
+		auth, err := checkStringListInMap(group, "authentication", "field", nil)
+		if err != nil {
+			return fmt.Errorf(`%s: "authentication" %w`, errPrefix, err)
+		}
+		if auth == nil {
```
[the remark this answers is not in the document]
Comment (st3v3nmw) [reply asserts/confdb_asserts.go]: The check ends up happening inside `registry/confdb_control/(AddGroup -> convertToAuthenticationMethod -> isValidAuthenticationMethod)`. It's a bit nested but if we do it here, we'll end up running the check twice. I could bump up the check though

File: features/features.go [hunk 2]
```diff
@@ -150,6 +154,7 @@ var featuresExported = map[SnapdFeature]bool{
 
 	RefreshAppAwarenessUX: true,
 	Registries:            true,
+	ConfdbControl:         true,
```
[the remark this answers is not in the document]
Comment (st3v3nmw) [reply features/features.go]: Yep, that flag shouldn't have been exported. Removed.

File: confdb/confdb_control.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+type Group struct {
+	Authentication []AuthenticationMethod
+	Views          []string
+}
+
+// AddGroup adds the group to an operator under the given authentication.
+func (op *Operator) AddGroup(views []string, auth []string) error {
+	if len(auth) == 0 {
+		return fmt.Errorf(`"authentication" must be a non-empty list`)
+	}
```
[the remark this answers is not in the document]
Comment (st3v3nmw) [reply confdb/confdb_control.go]: I've changed the errors to `cannot add group: "auth" must be a non-empty list` & `cannot add group: "views" must be a non-empty list` to be more descriptive. `auth` & `views` are the parameters passed to `op.AddGroup`

File: registry/confdb_control.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Operator holds the delegations for a single operator.
+type Operator struct {
+	ID     string
+	Groups []*Group
+}
+
+// Group holds a set of views delegated through the given authentication.
+type Group struct {
```
Comment (miguelpires) [inline registry/confdb_control.go:77]: Wondering whether `DelegatedViews` or `ControlGroup` wouldn't be better as Group is fairly generic but it's minor anyway

**Review by miguelpires — DISMISSED** [review]
thank you

File: registry/confdb_control.go (hunk 4 above)
Comment (st3v3nmw) [reply registry/confdb_control.go:77]: I was trying to match the spec but I agree, something like `ControlGroup` is clearer.

Thanks for the review!

Comment (pedronis) [reply registry/confdb_control.go:77]: given that is shows up as registry.Group I think we definitely need a more precise name, group in the context of the assertion is fine is it is scoped to the assertion, but the type name here is not

Comment (st3v3nmw) [reply registry/confdb_control.go]: @pedronis, I've renamed it to `ControlGroup`... `registry.ControlGroup` or `confdb.ControlGroup` seems better.

File: registry/registry_test.go [hunk 5]
```diff
@@ -142,7 +142,7 @@ func (*viewSuite) TestNewRegistry(c *C) {
 			c.Assert(err.Error(), Equals, tc.err, cmt)
 		} else {
 			c.Assert(err, IsNil, cmt)
-			c.Check(registry, Not(IsNil), cmt)
+			c.Check(registry, NotNil, cmt)
```
Comment (pedronis) [inline registry/registry_test.go:145]: for next time: it's best to do this kind of small tweaks/fixes as their own PR or commit with something like "drive-by" in the commit message

File: registry/registry.go [hunk 6]
```diff
@@ -450,7 +450,7 @@ func getPlaceholders(viewStr string) map[string]bool {
 	return placeholders
 }
 
-// View returns an view from the registry.
+// View returns a view from the registry.
```
Comment (pedronis) [inline registry/registry.go:453]: same

File: registry/confdb_control.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	"regexp"
+	"sort"
+	"strings"
+)
+
+var (
+	validAccountID = regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$")
+)
+
+type AuthenticationMethod string
```
Comment (pedronis) [inline registry/confdb_control.go:34]: this probably merits a doc comment

File: registry/confdb_control.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+// Operator holds the delegations for a single operator.
+type Operator struct {
+	ID     string
+	Groups []*ControlGroup
+}
+
+// ControlGroup holds a set of views delegated through the given authentication.
+type ControlGroup struct {
+	Authentication []AuthenticationMethod
+	Views          []string
```
Comment (pedronis) [inline registry/confdb_control.go:79]: shouldn't this contain structs already parsed into  account/confdb/view ?

File: registry/confdb_control.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		Views:          views,
+	}
+	op.Groups = append(op.Groups, group)
+
+	return nil
+}
+
+// compact replaces consecutive runs of equal elements with a single copy.
+// The provided slice s should be sorted.
+func compact[T comparable](s []T) []T {
```
Comment (pedronis) [inline registry/confdb_control.go:127]: unique might be a more common name for this

File: asserts/registry.go [hunk 10]
```diff
@@ -108,3 +109,110 @@ func assembleRegistry(assert assertionBase) (Assertion, error) {
 		timestamp:     timestamp,
 	}, nil
 }
+
+// ConfdbControl holds a confdb-control assertion, which holds lists of
+// views delegated by the device to an operator.
```
Comment (pedronis) [inline asserts/registry.go:114]: s/to an operator/to operators/

File: registry/confdb_control.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	switch AuthenticationMethod(value) {
+	case OperatorKey, Store:
+		return true
+	default:
+		return false
+	}
+}
+
+// convertToAuthenticationMethod converts and validates a []string to []AuthenticationMethod
+func convertToAuthenticationMethod(methods []string) ([]AuthenticationMethod, error) {
```
Comment (pedronis) [inline registry/confdb_control.go:52]: maybe "...Methods" as it is producing a slice

File: registry/confdb_control.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	sort.Slice(methods, func(i, j int) bool {
+		return methods[i] < methods[j]
+	})
+
+	// remove duplicates
+	methods = compact(methods)
```
Comment (pedronis) [inline registry/confdb_control.go:58]: I don't see code in the tests that check that this is done

**Review by pedronis — COMMENTED** [review]
thank you, did a pass, main comment is how we store views in ControlGroup for usage

File: registry/confdb_control.go (hunk 8 above)
Comment (st3v3nmw) [reply registry/confdb_control.go:79]: In the assertion, the list of groups is flat.. like so:

```yaml
groups:
  - operator-id:        john
    authentication: 	[ operator-key, store ]
    views:
      - a/b/c
      - d/e/f
  - operator-id:        john
    authentication: 	[ store ]
    views:
      - x/y/z
```

Internally though, we're grouping by operator to make this easier to look up later since all the API operations revolve around an operator. For instance, `isDelegated(operator, view)`, `delegate(operator, views, authMethods)`, `revoke(operator, ...)`.

```yaml
operators:
  john:
    ID: john
    Groups:
      - Authentication: [ operator-key, store ]
        Views:
          - a/b/c
          - d/e/f
      - Authentication: [ store ]
        Views:
          - x/y/z
```

Edit:

Or do you mean we should store the `Views` as a "set" for faster lookup.. Something like `map[string]struct{}`?

File: registry/confdb_control.go (hunk 12 above)
Comment (st3v3nmw) [reply registry/confdb_control.go]: It's a bit hidden but [this test](https://github.com/canonical/snapd/blob/15324821bcb5d36d5caa522faaf97df2f338d0bf/registry/confdb_control_test.go#L39) checks that behavior

File: registry/registry_test.go (hunk 5 above)
Comment (st3v3nmw) [reply registry/registry_test.go:145]: Thanks, I'll keep this in mind for future changes.

File: registry/confdb_control.go (hunk 7 above)
Comment (st3v3nmw) [reply registry/confdb_control.go:36]: I've added a comment here and to the `OperatorKey` & `Store` options

File: registry/confdb_control.go (hunk 8 above)
Comment (pedronis) [reply registry/confdb_control.go:83]: I mean of doing something like this after the parsing:

```
Views []*ViewRef

type ViewRef struct {
   Account string
   Confdb string
   View string
}
```
or similar

does this make sense to you @miguelpires as well ?

Comment (st3v3nmw) [reply registry/confdb_control.go]: Ah, okay. That makes sense. I've made the necessary changes to reflect this.

**Review by miguelpires — APPROVED** [review]
thanks. Looks good

**Review by pedronis — APPROVED** [review]
thank you

**Comment by st3v3nmw** [comment]
The failing spread tests are unrelated to the changes:

- google-arm:ubuntu-core-22-arm-64:tests/core/gadget-update-pc: _Error restoring: EOF_
- google-core:ubuntu-core-20-64:tests/core/mem-cgroup-disabled: _Error restoring: EOF_
- google-core:ubuntu-core-18-64:tests/core/kernel-base-gadget-pair-single-reboot:gadget_base: _Error restoring: EOF_
- google-core:ubuntu-core-18-64:tests/core/kernel-base-gadget-single-reboot: _cannot obtain the remote system boot_id: ssh: unexpected packet in response to channel open: <nil>_
- google:ubuntu-25.04-64:tests/main/apparmor-prompting-integration-tests:create_multiple_allow: _oom-killer got invoked during the tests_
- google:ubuntu-25.04-64:tests/main/apparmor-prompting-integration-tests:download_file_defaults: _oom-killer got invoked during the tests_
- google:ubuntu-24.10-64:tests/main/aux-info: _oom-killer got invoked during the tests_
- google:ubuntu-24.04-64:tests/main/document-interfaces-url: _The following interfaces do not have a documentation webpage:  physical-memory-observe_
- google:ubuntu-22.04-64:tests/main/snapd-state: _Time 2024-12-09T09:04:13+00:00 too far from current time (2024-12-09T10:36:42+00:00)... Probably the test took too long to run?_
- google:ubuntu-16.04-64:tests/main/degraded: _? unrelated_
- openstack:opensuse-15.5-64:tests/completion/: _failed suite prepare... too early for operation, device not yet seeded or device model not acknowledged_
- openstack:fedora-40-64:tests/main/: _failed suite prepare... too early for operation, device not yet seeded or device model not acknowledged_
